torch/distributed/_spmd/api.py

```diff
@@ -51,3 +76,208 @@ def forward(self, *args: Tuple[object], **kwargs: Dict[str, object]) -> object:
+            type to ``Override`` implementation. The ``Override`` instance
+            provides an ``nn.Module`` replacement during tracing and a graph
+            transformation function after tracing.
+    """
+
+    def inner(func: Callable):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            # 1. Extract nn.Module and Optimizer from args and kwargs
+            # FIXME(@mrshenli): support multiple nn.Module instances
+            # FIXME(@mrshenli): support multiple Optiimzer instances
+            mod, opt = None, None
+            for arg in pytree.tree_flatten(list(args) + list(kwargs.values()))[
+                0
+            ]:
+                if isinstance(arg, nn.Module):
+                    assert mod is None, "Only support single nn.Module for now"
+                    mod = arg
+                if isinstance(arg, torch.optim.Optimizer):
+                    assert opt is None, "Only support single Optimizer for now"
+                    opt = arg
+
+            assert mod is not None and opt is not None
+
+            # 2. Override target submodules (e.g., MoE) with dummy replacements
+            if override_map:
+                accessor = NamedMemberAccessor(mod)
+
+                for typ, override in override_map.items():
+                    for name, submodule in mod.named_modules():
+                        if isinstance(submodule, typ):
+                            accessor.swap_submodule(
+                                name, override.replacement(submodule)
+                            )
+
+            # 3. Trace statelss version of the train_step
+            params_and_buffers = {
+                **dict(mod.named_parameters(remove_duplicate=False)),
+                **dict(mod.named_buffers(remove_duplicate=False)),
+            }
+
+            opt_states, spec = pytree.tree_flatten(dict(opt.state))
+            named_states = {}
+            for n, p in params_and_buffers.items():
+                if p in opt.state:
+                    named_states[n] = opt.state[p]
+
+            def stateless_func(
+                func, args, kwargs, named_states, params_and_buffers
+            ):
+                with stateless._reparametrize_module(
+                    mod, params_and_buffers
+                ), _rematerialize_optimizer(
+                    opt, named_states, params_and_buffers
+                ):
+                    ret = func(*args, **kwargs)
+                    # make sure updated parameters are returned
+                    return ret, list(mod.parameters())
+
+            # parameters are at the end of placeholders
+            # FIXME(@mrshenli): let make_fx ignore nn.Module in args? or it should
+            # eliminate unused placeholders?
+            # FIXME: Using symbolic tracing to work around. Otherwise it hits
+            # shape mismatch error, as we use local inputs to trace local graph
```

Comment:
Yep, I tried that but does not seem to work. Maybe I didn't do it correctly. IIUC, the problem I am hitting is that:

*  if I use fake tensor with global batch size, operators like `expand` will embed/hardcode the fake tensor shape as its argument in the graph. But during local execution (before DTensor expansion), it is actually one shard of that, so there will be shape mismatch with the hardcoded one. That expansion operation is from the loss function (e.g., `sum()`). I think we didn't hit this problem in SPMD ([here](https://github.com/pytorch/pytorch/blob/17567e5b29a77dee137dfc365e9ed3f1cf358950/torch/distributed/_spmd/distribute.py#L612)) because SPMD does not include the loss function.
*  If I use local batch size, then the expand operator will hardcode local batch size in the graph. Later during DTensor expansion, the FakeTensor wrapping DTensor will have the global batch size, and hence mismatch.

Not 100% sure if the above understanding is correct.